gcs/blockcf2/blockcf.go

```diff
@@ -0,0 +1,395 @@
+// associated version keyed by an outpoint.  It is used within this package as a
+// generic means to provide the scripts referenced by all of the inputs to
+// transactions within a block that are needed to construct a filter.  The
+// boolean return indicates whether or not the script and version for the
+// provided outpoint was found.
+type PrevScripter interface {
+	PrevScript(*wire.OutPoint) (uint16, []byte, bool)
+}
+
+// Regular builds a GCS filter from a block and the previous output scripts it
+// references as inputs.  The filter will be keyed by the merkle root of the
+// block.
+//
+// The following section describes the items that will be added to the filter,
+// however, there are a few special cases that apply:
+//
+// - Scripts that are not version 0, empty, or otherwise provably unspendable
+//   are NOT included
+// - Output scripts for transactions in the stake tree do NOT include the
+//   initial stake opcode tag (OP_SS*)
+//   - This allows users of the filter to only match against a normal P2PKH or
+//     P2SH script, instead of many extra matches for each tag
+//
+// Considering the aforementioned exceptions, the filter will contain the
+// following items for transactions in the regular tree:
+// - Previous output scripts referenced by the transaction inputs with the
+//   exception of the coinbase
+// - Output scripts in the transaction outputs
+//
+// In addition, also considering the aforementioned exceptions, the filter will
+// contain the following items for transactions in the stake tree:
+// - For ticket purchases:
+//   - Previous output scripts referenced by the transaction inputs
+//   - Commitment output scripts (converted from the commitment hash and type)
+//   - Change output scripts except those that are provably unspendable
+// - For votes:
+//   - Subsidy generation output scripts
+//   - Output scripts that pay the original ticket commitments
+// - For revocations:
+//   - Output scripts that pay the original ticket commitments
+func Regular(block *wire.MsgBlock, prevScripts PrevScripter) (*gcs.FilterV2, error) {
+	// There will typically be data entries for at least one output and one
+	// input per regular transaction in the block, excepting the coinbase, and
+	// an average of two per stake transaction, though stake transactions vary
+	// wildly.
+	//
+	// In practice, most transactions will have more inputs and outputs than
+	// this value, but pre-allocate enough space in the backing array for a
+	// reasonable minimum value to reduce the number of allocations.
+	numEntriesHint := len(block.Transactions)*2 + len(block.STransactions)
+	data := make(Entries, 0, numEntriesHint)
+
+	// For regular transactions, add all referenced previous output scripts,
+	// except the coinbase, and all output scripts.
+	for i, tx := range block.Transactions {
+		for _, txOut := range tx.TxOut {
+			data.AddRegularPkScript(txOut.PkScript)
```

Comment:
Yes, the intent is to allow the querying of `OP_RETURN`s.  However, you're right that it really should be excluding the other conditions as well, most notably non-zero script versions.  I'll think about cleanest way to handle this a bit, but I suspect updating the `excludeFromFilter` function to take a flag whether or not exclude `OP_RETURN`s is probably the best way.

In reality, there should never be an input to a previous output that is an `OP_RETURN` because the block would fail validation due to attempting to spend an unspendable output, so I could probably just remove that check, but I think it's prudent to be overly explicit since these will now be enforced by consensus.

Reply:
Alternatively, we can simply exclude all `OP_RETURN`s from the filters in general and just call the function here as it is.  It could theoretically be useful to be able to query for random `OP_RETURN`s in some niche cases, which was the intent, however, it does come at the cost of increasing the size of the filters slightly, so I'm not entirely convinced they should be included.

Reply:
As a point of consideration, here is the difference between the total size of filters up to block 382281 created with the `OP_RETURN` vs without:

```
total filter bytes with OP_RETURNs: 71066500 (67.8 MiB)
total filter bytes without OP_RETURNs: 54640705 (52.1 MiB)
```

Reply:
I've added the appropriate check.

For reference, we've discussed this via real-time chat and decided to include `OP_RETURN`s in the filters since they don't consume a significant amount of extra space, there are limited number allowed per transaction, and it would be a significant amount of work to modify consensus in the future to include them.

Also, after further discussion, we've decided to rely on the fact that will never be an input to a previous output that is an `OP_RETURN` because the block would fail validation due to attempting to spend an unspendable output versus adding an additional flag to check the impossible condition.  The primary reasoning for this is that and any modifications that would change the semantics here would necessarily be a change to consensus and require examining the filters as a part of that change anyway.